backends/dlc/scummvmcloud.cpp

```diff
@@ -0,0 +1,213 @@
+		removeCacheFile(relativeFilePath);
+		// add downloaded game entry in scummvm configuration file
+		addEntryToConfig(destPath);
+		// handle next download
+		DLCMan._queuedDownloadTasks.front()->state = DLCDesc::kDownloaded;
+		DLCMan.refreshDLCList();
+		DLCMan._queuedDownloadTasks.pop();
+		DLCMan._dlcsInProgress.remove_at(0);
+		DLCMan.processDownloadQueue();
+	}
+}
+
+void ScummVMCloud::errorCallback(Networking::ErrorResponse error) {
+	// error downloading - start next download in queue
+	DLCMan._queuedDownloadTasks.front()->state = DLCDesc::kErrorDownloading;
+	DLCMan._queuedDownloadTasks.pop();
+	DLCMan._dlcsInProgress.remove_at(0);
+	DLCMan.processDownloadQueue();
+}
+
+void ScummVMCloud::startDownloadAsync(const Common::String &id, const Common::String &url) {
+	Common::String localFile = normalizePath(ConfMan.get("dlcspath") + "/" + id, '/');
+
+	_rq = new Networking::SessionRequest(url, localFile,
+		new Common::Callback<ScummVMCloud, Networking::DataResponse>(this, &ScummVMCloud::downloadFileCallback),
+		new Common::Callback<ScummVMCloud, Networking::ErrorResponse>(this, &ScummVMCloud::errorCallback));
+
+	_rq->start();
+}
+
+void ScummVMCloud::extractZip(const Common::Path &file, const Common::Path &destPath) {
+	Common::Archive *dataArchive = nullptr;
+	Common::Path dlcPath = Common::Path(ConfMan.get("dlcspath"));
+	Common::FSNode *fs = new Common::FSNode(dlcPath.join(file));
+	if (fs->exists()) {
+		dataArchive = Common::makeZipArchive(*fs);
+		// dataArchive is nullptr if zip file is incomplete
+		if (dataArchive != nullptr) {
+			dataArchive->dumpArchive(destPath.toString());
+			DLCMan._errorText = "";
+		} else {
+			DLCMan._errorText = "ERROR: " + DLCMan._queuedDownloadTasks.front()->name + ": Archive is broken, please re-download!";
+		}
+	}
+	delete fs;
+	delete dataArchive;
+}
+
+void ScummVMCloud::removeCacheFile(const Common::Path &file) {
+	Common::Path dlcPath = Common::Path(ConfMan.get("dlcspath"));
+	Common::Path fileToDelete = dlcPath.join(file);
+#if defined(POSIX)
+	unlink(fileToDelete.toString().c_str());
+#elif defined(WIN32)
+	_unlink(fileToDelete.toString().c_str());
+#else
+	warning("ScummVMCloud::removeCacheFile(): Removing is unimplemented");
+#endif
+}
+
+void ScummVMCloud::addEntryToConfig(Common::Path gamePath) {
+	Common::FSNode dir(gamePath);
+	Common::FSList fsnodes;
+	if (!dir.getChildren(fsnodes, Common::FSNode::kListAll)) {
+		warning("ScummVMCloud::addEntryToConfig(): Game directory does not exists");
+		return;
+	}
+	if (fsnodes.size() == 1 && fsnodes[0].isDirectory()) {
+		// if extraction process created a new folder inside gamePath, set gamePath to that directory
+		gamePath = gamePath.appendComponent(fsnodes[0].getFileName());
+	}
+	// add a new entry in scummvm config file
+	Common::String domain = EngineMan.generateUniqueDomain(DLCMan._queuedDownloadTasks.front()->gameid);
+	ConfMan.addGameDomain(domain);
+	ConfMan.set("engineid", DLCMan._queuedDownloadTasks.front()->engineid, domain);
```

Comment:
For brevity, please put `DLCMan._queuedDownloadTasks.front()` into a local variable and reuse